File: engines/temporal/internal/corpus/probe_test.go
package corpus_test

import (
	"context"
	"errors"
	"sync"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/controller-agent/temporal-engine/internal/corpus"
)

// fakeProber answers per (connection, source), counts the probes it was
// actually asked to make, and can be switched to connection granularity.
type fakeProber struct {
	granularity corpus.Granularity
	results     map[string]corpus.ProbeResult
	errs        map[string]error

	mu     sync.Mutex
	probes []corpus.ProbeRequest
}

func (f *fakeProber) Granularity(string) corpus.Granularity {
	if f.granularity == "" {
		return corpus.GranularityResource
	}
	return f.granularity
}

func (f *fakeProber) Probe(_ context.Context, req corpus.ProbeRequest) (corpus.ProbeResult, error) {
	f.mu.Lock()
	f.probes = append(f.probes, req)
	f.mu.Unlock()

	key := req.CorpusID + "/" + req.SourceID
	if err, ok := f.errs[key]; ok {
		return corpus.ProbeResult{}, err
	}
	if result, ok := f.results[key]; ok {
		return result, nil
	}
	return corpus.ProbeResult{Allowed: false}, nil
}

func (f *fakeProber) probeCount() int {
	f.mu.Lock()
	defer f.mu.Unlock()
	return len(f.probes)
}

func candidate(connectionID, sourceID, version string, score float32) corpus.Hit {
	return corpus.Hit{
		Score: score,
		Chunk: corpus.Chunk{
			CorpusID: connectionID,
			SourceID: sourceID,
			Version:  version,
			// Deliberately misleading mirror metadata: Authorize must never let
			// these reach a citation.
			Title:       "STALE MIRROR TITLE",
			SourceURL:   "https://mirror.invalid/leaked",
			ContentHash: connectionID + sourceID + version,
			Text:        "…passage…",
		},
	}
}

func TestAuthorizeTakesCitationFieldsFromTheProbeNotTheMirror(t *testing.T) {
	prober := &fakeProber{results: map[string]corpus.ProbeResult{
		"snc-confluence/page-1": {
			Allowed: true,
			Title:   "Auth design",
			URL:     "https://example.atlassian.net/wiki/page-1",
			Version: "v7",
		},
	}}

	outcome, err := corpus.Authorize(context.Background(), prober,
		[]corpus.Hit{candidate("snc-confluence", "page-1", "v7", 0.9)})

	require.NoError(t, err)
	require.Len(t, outcome.Chunks, 1)
	// Citations are content: a title or URL taken from the mirror would bypass
	// the whole design at the last step.
	require.Equal(t, "Auth design", outcome.Chunks[0].Title)
	require.Equal(t, "https://example.atlassian.net/wiki/page-1", outcome.Chunks[0].URL)
	require.NotContains(t, outcome.Chunks[0].URL, "mirror.invalid")
}

func TestAuthorizeDropsWhatTheSourceRefuses(t *testing.T) {
	prober := &fakeProber{
		results: map[string]corpus.ProbeResult{
			"snc-confluence/page-1": {Allowed: true, Title: "Readable", URL: "u", Version: "v1"},
		},
		errs: map[string]error{
			"snc-confluence/page-2": &corpus.PermissionDenied{Err: errors.New("403")},
		},
	}

	outcome, err := corpus.Authorize(context.Background(), prober, []corpus.Hit{
		candidate("snc-confluence", "page-1", "v1", 0.9),
		candidate("snc-confluence", "page-2", "v1", 0.8),
	})

	require.NoError(t, err)
	require.Len(t, outcome.Chunks, 1)
	require.Equal(t, 1, outcome.Denied)
	require.Empty(t, outcome.Undetermined)
}

func TestAuthorizeTreatsAllowedFalseAsADenial(t *testing.T) {
	prober := &fakeProber{results: map[string]corpus.ProbeResult{
		"c/s": {Allowed: false},
	}}

	outcome, err := corpus.Authorize(context.Background(), prober,
		[]corpus.Hit{candidate("c", "s", "v1", 0.5)})

	require.NoError(t, err)
	require.Empty(t, outcome.Chunks)
	require.Equal(t, 1, outcome.Denied)
}

func TestAuthorizeDoesNotSilentlyDropOnATransientFailure(t *testing.T) {
	prober := &fakeProber{
		results: map[string]corpus.ProbeResult{
			"c/ok": {Allowed: true, Title: "t", URL: "u", Version: "v1"},
		},
		errs: map[string]error{
			"c/busy": &corpus.Transient{Err: errors.New("429 rate limited")},
		},
	}

	outcome, err := corpus.Authorize(context.Background(), prober, []corpus.Hit{
		candidate("c", "ok", "v1", 0.9),
		candidate("c", "busy", "v1", 0.8),
	})

	require.NoError(t, err)
	require.Len(t, outcome.Chunks, 1)
	// A 429 is not a denial. Counting it as one would quietly shrink the answer
	// and make the same question return different evidence on a retry.
	require.Zero(t, outcome.Denied)
	require.Equal(t, []string{"c/busy"}, outcome.Undetermined)
}

func TestAuthorizeFailsOnAnUnclassifiedDriverError(t *testing.T) {
	prober := &fakeProber{errs: map[string]error{
		"c/s": errors.New("driver forgot to classify this"),
	}}

	_, err := corpus.Authorize(context.Background(), prober,
		[]corpus.Hit{candidate("c", "s", "v1", 0.5)})

	// Guessing whether an unclassified error meant "denied" or "busy" is how a
	// leak gets introduced, so it fails the search instead.
	require.Error(t, err)
}

func TestAuthorizeDeduplicatesProbesPerSource(t *testing.T) {
	prober := &fakeProber{results: map[string]corpus.ProbeResult{
		"c/page-1": {Allowed: true, Title: "t", URL: "u", Version: "v1"},
	}}

	hits := make([]corpus.Hit, 0, 8)
	for i := 0; i < 8; i++ {
		hit := candidate("c", "page-1", "v1", float32(i)/10)
		hit.Chunk.ContentHash = string(rune('a' + i)) // eight chunks, one page
		hits = append(hits, hit)
	}

	outcome, err := corpus.Authorize(context.Background(), prober, hits)

	require.NoError(t, err)
	require.Len(t, outcome.Chunks, 8)
	require.Equal(t, 1, prober.probeCount(),
		"eight chunks from one page must cost one probe, not eight")
}

func TestAuthorizeProbesOncePerConnectionForChannelScopedProviders(t *testing.T) {
	// Slack authorizes a CHANNEL: membership is the access unit, so one probe
	// settles every candidate from that connection.
	prober := &fakeProber{
		granularity: corpus.GranularityConnection,
		results: map[string]corpus.ProbeResult{
			"snc-slack-eng/": {Allowed: true, Title: "#snc-eng", URL: "u", Version: ""},
		},
	}

	outcome, err := corpus.Authorize(context.Background(), prober, []corpus.Hit{
		candidate("snc-slack-eng", "msg-1", "", 0.9),
		candidate("snc-slack-eng", "msg-2", "", 0.8),
		candidate("snc-slack-eng", "msg-3", "", 0.7),
	})

	require.NoError(t, err)
	require.Len(t, outcome.Chunks, 3)
	require.Equal(t, 1, prober.probeCount())
}

func TestAuthorizeMarksStaleWhenTheSourceHasMovedOn(t *testing.T) {
	prober := &fakeProber{results: map[string]corpus.ProbeResult{
		"c/page-1": {Allowed: true, Title: "t", URL: "u", Version: "v9"},
		"c/page-2": {Allowed: true, Title: "t", URL: "u", Version: "v2"},
		"c/page-3": {Allowed: true, Title: "t", URL: "u", Version: ""},
	}}

	outcome, err := corpus.Authorize(context.Background(), prober, []corpus.Hit{
		candidate("c", "page-1", "v7", 0.9), // indexed at v7, source at v9
		candidate("c", "page-2", "v2", 0.8), // current
		candidate("c", "page-3", "v1", 0.7), // provider reports no version
	})

	require.NoError(t, err)
	require.Len(t, outcome.Chunks, 3)
	require.True(t, outcome.Chunks[0].Stale, "readable but overtaken — not dropped, just old")
	require.False(t, outcome.Chunks[1].Stale)
	require.False(t, outcome.Chunks[2].Stale,
		"a provider that reports no version cannot be shown to be stale, so do not claim it is")
}

func TestAuthorizePreservesRankOrder(t *testing.T) {
	prober := &fakeProber{results: map[string]corpus.ProbeResult{
		"c/a": {Allowed: true, Title: "a", URL: "ua", Version: "v1"},
		"c/b": {Allowed: true, Title: "b", URL: "ub", Version: "v1"},
		"c/c": {Allowed: true, Title: "c", URL: "uc", Version: "v1"},
	}}

	outcome, err := corpus.Authorize(context.Background(), prober, []corpus.Hit{
		candidate("c", "a", "v1", 0.9),
		candidate("c", "b", "v1", 0.5),
		candidate("c", "c", "v1", 0.1),
	})

	require.NoError(t, err)
	require.Equal(t, []string{"a", "b", "c"},
		[]string{outcome.Chunks[0].Title, outcome.Chunks[1].Title, outcome.Chunks[2].Title})
}

func TestAuthorizeWithNoCandidates(t *testing.T) {
	outcome, err := corpus.Authorize(context.Background(), &fakeProber{}, nil)
	require.NoError(t, err)
	require.Empty(t, outcome.Chunks)
}

File: apps/agent-orchestrator/src/knowledge-base/probe.test.ts
import { describe, expect, it } from "vitest";
import {
  authorize,
  PermissionDeniedError,
  TransientProbeError,
  type Granularity,
  type ProbeRequest,
  type ProbeResult,
  type Prober,
} from "./probe.js";
import type { CorpusSearchResult } from "./types.js";

/**
 * Answers per `connection/source`, counts the probes it was actually asked to
 * make, and can be switched to connection granularity.
 */
class FakeProber implements Prober {
  readonly probes: ProbeRequest[] = [];

  constructor(
    private readonly results: Record<string, ProbeResult> = {},
    private readonly errors: Record<string, Error> = {},
    private readonly unit: Granularity = "resource",
  ) {}

  granularity(): Granularity {
    return this.unit;
  }

  async probe(request: ProbeRequest): Promise<ProbeResult> {
    this.probes.push(request);
    const key = `${request.connectionId}/${request.sourceId ?? ""}`;
    const failure = this.errors[key];
    if (failure) throw failure;
    return this.results[key] ?? { allowed: false, title: "", url: "" };
  }
}

function candidate(
  connectionId: string,
  sourceId: string,
  version: string,
  score: number,
): CorpusSearchResult {
  return {
    score,
    chunk: {
      connectionId,
      sourceId,
      version,
      // Deliberately misleading mirror metadata: authorize must never let these
      // reach a citation.
      title: "STALE MIRROR TITLE",
      sourceUrl: "https://mirror.invalid/leaked",
      contentHash: `${connectionId}${sourceId}${version}`,
      text: "…passage…",
    },
  };
}

describe("authorize", () => {
  it("takes citation fields from the probe, not the mirror", async () => {
    const prober = new FakeProber({
      "globex-confluence/page-1": {
        allowed: true,
        title: "Auth design",
        url: "https://example.atlassian.net/wiki/page-1",
        version: "v7",
      },
    });

    const outcome = await authorize(prober, [candidate("globex-confluence", "page-1", "v7", 0.9)]);

    expect(outcome.chunks).toHaveLength(1);
    // Citations are content: a title or URL from the mirror would bypass the
    // whole design at the last step.
    expect(outcome.chunks[0].title).toBe("Auth design");
    expect(outcome.chunks[0].url).not.toContain("mirror.invalid");
  });

  it("drops what the source refuses", async () => {
    const prober = new FakeProber(
      { "globex-confluence/page-1": { allowed: true, title: "Readable", url: "u", version: "v1" } },
      { "globex-confluence/page-2": new PermissionDeniedError("403") },
    );

    const outcome = await authorize(prober, [
      candidate("globex-confluence", "page-1", "v1", 0.9),
      candidate("globex-confluence", "page-2", "v1", 0.8),
    ]);

    expect(outcome.chunks).toHaveLength(1);
    expect(outcome.denied).toBe(1);
    expect(outcome.undetermined).toEqual([]);
  });

  it("treats allowed:false as a denial", async () => {
    const outcome = await authorize(new FakeProber({ "c/s": { allowed: false, title: "", url: "" } }), [
      candidate("c", "s", "v1", 0.5),
    ]);

    expect(outcome.chunks).toEqual([]);
    expect(outcome.denied).toBe(1);
  });

  it("does not silently drop on a transient failure", async () => {
    const prober = new FakeProber(
      { "c/ok": { allowed: true, title: "t", url: "u", version: "v1" } },
      { "c/busy": new TransientProbeError("429 rate limited") },
    );

    const outcome = await authorize(prober, [
      candidate("c", "ok", "v1", 0.9),
      candidate("c", "busy", "v1", 0.8),
    ]);

    expect(outcome.chunks).toHaveLength(1);
    // A 429 is not a denial. Counting it as one would quietly shrink the answer
    // and make the same question return different evidence on a retry.
    expect(outcome.denied).toBe(0);
    expect(outcome.undetermined).toEqual(["c/busy"]);
  });

  it("fails on an unclassified driver error", async () => {
    const prober = new FakeProber({}, { "c/s": new Error("driver forgot to classify this") });

    // Guessing whether an unclassified error meant "denied" or "busy" is how a
    // leak gets introduced, so it fails the search instead.
    await expect(authorize(prober, [candidate("c", "s", "v1", 0.5)])).rejects.toThrow(
      /forgot to classify/,
    );
  });

  it("de-duplicates probes per source", async () => {
    const prober = new FakeProber({
      "c/page-1": { allowed: true, title: "t", url: "u", version: "v1" },
    });

    const hits = Array.from({ length: 8 }, (_, i) => {
      const hit = candidate("c", "page-1", "v1", i / 10);
      hit.chunk.contentHash = `chunk-${i}`; // eight chunks, one page
      return hit;
    });

    const outcome = await authorize(prober, hits);

    expect(outcome.chunks).toHaveLength(8);
    expect(prober.probes).toHaveLength(1);
  });

  it("probes once per connection for channel-scoped providers", async () => {
    // Slack authorizes a CHANNEL: membership is the access unit, so one probe
    // settles every candidate from that connection.
    const prober = new FakeProber(
      { "globex-slack-eng/": { allowed: true, title: "#globex-eng", url: "u" } },
      {},
      "connection",
    );

    const outcome = await authorize(prober, [
      candidate("globex-slack-eng", "msg-1", "", 0.9),
      candidate("globex-slack-eng", "msg-2", "", 0.8),
      candidate("globex-slack-eng", "msg-3", "", 0.7),
    ]);

    expect(outcome.chunks).toHaveLength(3);
    expect(prober.probes).toHaveLength(1);
  });

  it("marks stale when the source has moved on", async () => {
    const prober = new FakeProber({
      "c/page-1": { allowed: true, title: "t", url: "u", version: "v9" },
      "c/page-2": { allowed: true, title: "t", url: "u", version: "v2" },
      "c/page-3": { allowed: true, title: "t", url: "u" },
    });

    const outcome = await authorize(prober, [
      candidate("c", "page-1", "v7", 0.9), // indexed at v7, source at v9
      candidate("c", "page-2", "v2", 0.8), // current
      candidate("c", "page-3", "v1", 0.7), // provider reports no version
    ]);

    expect(outcome.chunks[0].stale).toBe(true);
    expect(outcome.chunks[1].stale).toBe(false);
    // A provider reporting no version cannot show a chunk to be stale.
    expect(outcome.chunks[2].stale).toBe(false);
  });

  it("preserves rank order", async () => {
    const prober = new FakeProber({
      "c/a": { allowed: true, title: "a", url: "ua", version: "v1" },
      "c/b": { allowed: true, title: "b", url: "ub", version: "v1" },
      "c/c": { allowed: true, title: "c", url: "uc", version: "v1" },
    });

    const outcome = await authorize(prober, [
      candidate("c", "a", "v1", 0.9),
      candidate("c", "b", "v1", 0.5),
      candidate("c", "c", "v1", 0.1),
    ]);

    expect(outcome.chunks.map((c) => c.title)).toEqual(["a", "b", "c"]);
  });

  it("handles no candidates", async () => {
    const outcome = await authorize(new FakeProber(), []);
    expect(outcome.chunks).toEqual([]);
  });
});
